dataflow/gpu-examples/conftest.py

```diff
@@ -0,0 +1,508 @@
+            yield dict(row)
+
+    @staticmethod
+    def pubsub_topic(name: str, project: str = PROJECT) -> str:
+        from google.cloud import pubsub
+
+        publisher_client = pubsub.PublisherClient()
+        topic_path = publisher_client.topic_path(project, Utils.hyphen_name(name))
+        topic = publisher_client.create_topic(topic_path)
+
+        logging.info(f"Created pubsub_topic: {topic.name}")
+        yield topic.name
+
+        # Due to the pinned library dependencies in apache-beam, client
+        # library throws an error upon deletion.
+        # We use gcloud for a workaround. See also:
+        # https://github.com/GoogleCloudPlatform/python-docs-samples/issues/4492
+        cmd = ["gcloud", "pubsub", "--project", project, "topics", "delete", topic.name]
+        logging.info(f"{cmd}")
+        subprocess.run(cmd, check=True)
+        logging.info(f"Deleted pubsub_topic: {topic.name}")
+
+    @staticmethod
+    def pubsub_subscription(
+        topic_path: str,
+        name: str,
+        project: str = PROJECT,
+    ) -> str:
+        from google.cloud import pubsub
+
+        subscriber = pubsub.SubscriberClient()
+        subscription_path = subscriber.subscription_path(
+            project, Utils.hyphen_name(name)
+        )
+        subscription = subscriber.create_subscription(subscription_path, topic_path)
+
+        logging.info(f"Created pubsub_subscription: {subscription.name}")
+        yield subscription.name
+
+        # Due to the pinned library dependencies in apache-beam, client
+        # library throws an error upon deletion.
+        # We use gcloud for a workaround. See also:
+        # https://github.com/GoogleCloudPlatform/python-docs-samples/issues/4492
+        cmd = [
+            "gcloud",
+            "pubsub",
+            "--project",
+            project,
+            "subscriptions",
+            "delete",
+            subscription.name,
+        ]
+        logging.info(f"{cmd}")
+        subprocess.run(cmd, check=True)
+        logging.info(f"Deleted pubsub_subscription: {subscription.name}")
+
+    @staticmethod
+    def pubsub_publisher(
+        topic_path: str,
+        new_msg: Callable[[int], str] = lambda i: json.dumps(
+            {"id": i, "content": f"message {i}"}
+        ),
+        sleep_sec: int = 1,
+    ) -> bool:
+        from google.cloud import pubsub
+
+        def _infinite_publish_job() -> None:
+            publisher_client = pubsub.PublisherClient()
+            for i in itertools.count():
+                msg = new_msg(i)
+                publisher_client.publish(topic_path, msg.encode("utf-8")).result()
+                time.sleep(sleep_sec)
+
+        # Start a subprocess in the background to do the publishing.
+        logging.info(f"Starting publisher on {topic_path}")
+        p = mp.Process(target=_infinite_publish_job)
+        p.start()
+
+        yield p.is_alive()
+
+        # For cleanup, terminate the background process.
```

Comment:
Is there ever a circumstance where this process wouldn't be stopped?

Reply:
I made some experiments:

* Joining with timeout 0 actually doesn't do anything since it's an infinite loop so it never finishes.
* Terminate always stops the subprocess, but my experiments are linear, and pytest might be doing some nasty concurrency
* If I don't terminate the process, the whole thing hangs because the main process waits for the subprocess to finish, which might explain the intermittent hangs I was getting (if for some reason the fixture cleanup doesn't execute)
* Apparently, we can set up [`daemon`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.daemon) to True and the main process no longer waits for the subprocess to finish, so it terminates it automatically even if I don't explicitly terminate it :)

I deleted the `join`, kept the `terminate`, but also made the subprocess into a daemon, and added a large explanation on it.